### ISSUE TITLE
Domi: MDIterator and MDRevIterator unit tests failing in nightly 

### ISSUE THREAD

**Issue by william76** [trilinos/Trilinos#1326]
@wfspotz

In the [nightly Clean track](https://testing.sandia.gov/cdash/index.php?project=Trilinos##Clean), we're seeing 2 ctest failures:
- Domi_MDIterator_UnitTests
- Domi_MDRevIterator_UnitTests

This is occurring on the jobs:
- Linux-gcc-4.9.3-SERIAL_Release_gcc_4.9.3__DEV
- Linux-gcc-4.9.3-MPI_Release_gcc_4.9.3_openmpi_1.8.7_DEV
- Linux-GCC-4.7.2-SERIAL_OPT_DEV

From MDIterator:
```
The following tests FAILED:
    12. MDIterator_double_MDArrayIteratorIndex_UnitTest ... 
    13. MDIterator_double_MDArrayViewIteratorIndex_UnitTest ... 
    14. MDIterator_double_MDArrayRCPIteratorIndex_UnitTest ... 
    30. MDIterator_float_MDArrayIteratorIndex_UnitTest ... 
    31. MDIterator_float_MDArrayViewIteratorIndex_UnitTest ... 
    32. MDIterator_float_MDArrayRCPIteratorIndex_UnitTest ... 
    48. MDIterator_int_MDArrayIteratorIndex_UnitTest ... 
    49. MDIterator_int_MDArrayViewIteratorIndex_UnitTest ... 
    50. MDIterator_int_MDArrayRCPIteratorIndex_UnitTest ... 
    66. MDIterator_long_MDArrayIteratorIndex_UnitTest ... 
    67. MDIterator_long_MDArrayViewIteratorIndex_UnitTest ... 
    68. MDIterator_long_MDArrayRCPIteratorIndex_UnitTest ... 
```

From MDRevIterator:
```
The following tests FAILED:
    12. MDRevIterator_double_MDArrayRevIteratorIndex_UnitTest ... 
    13. MDRevIterator_double_MDArrayViewRevIteratorIndex_UnitTest ... 
    14. MDRevIterator_double_MDArrayRCPRevIteratorIndex_UnitTest ... 
    29. MDRevIterator_float_MDArrayRevIteratorIndex_UnitTest ... 
    30. MDRevIterator_float_MDArrayViewRevIteratorIndex_UnitTest ... 
    31. MDRevIterator_float_MDArrayRCPRevIteratorIndex_UnitTest ... 
    46. MDRevIterator_int_MDArrayRevIteratorIndex_UnitTest ... 
    47. MDRevIterator_int_MDArrayViewRevIteratorIndex_UnitTest ... 
    48. MDRevIterator_int_MDArrayRCPRevIteratorIndex_UnitTest ... 
    63. MDRevIterator_long_MDArrayRevIteratorIndex_UnitTest ... 
    64. MDRevIterator_long_MDArrayViewRevIteratorIndex_UnitTest ... 
    65. MDRevIterator_long_MDArrayRCPRevIteratorIndex_UnitTest ... 
``` 

During our stand-up meeting this morning, @bmpersc said he'd add the configuration options from these tests to the ticket.

**Comment by bmpersc** [comment]
@wfspotz I have been able to replicate this on my rhel6 machine using the sems modules  The configure I used is:

cmake \
  -DCMAKE_INSTALL_PREFIX=/scratch/bmpersc/sierra_integration2/test_runs/tests/trilinos_install \
  -DTPL_ENABLE_MPI:BOOL=ON \
    -DMPI_BASE_DIR=/projects/sems/install/rhel6-x86_64/sems/compiler/gcc/4.9.3/openmpi/1.8.7 \
  -D BUILD_SHARED_LIBS:BOOL=ON \
  -DTrilinos_ENABLE_STK:BOOL=ON \
  -DTrilinos_ENABLE_SEACAS:BOOL=ON \
  -DTrilinos_ENABLE_Epetra:BOOL=ON \
  -DTrilinos_ENABLE_Tpetra:BOOL=ON \
  -DTrilinos_ENABLE_ML:BOOL=ON \
  -DTrilinos_ENABLE_MueLu:BOOL=ON \
  -DTrilinos_ENABLE_EpetraExt:BOOL=ON \
  -DTrilinos_ENABLE_AztecOO:BOOL=ON \
  -DTrilinos_ENABLE_Belos:BOOL=ON \
  -DTrilinos_ENABLE_Ifpack2:BOOL=ON \
  -DTrilinos_ENABLE_Amesos2:BOOL=ON \
  -DTrilinos_ENABLE_Zoltan2:BOOL=ON \
  -DTrilinos_ENABLE_Ifpack:BOOL=ON \
  -DTrilinos_ENABLE_Amesos:BOOL=ON \
  -DTrilinos_ENABLE_Zoltan:BOOL=ON \
  -DTrilinos_ENABLE_PyTrilinos:BOOL=OFF \
  -DTrilinos_ENABLE_TESTS:BOOL=ON \
  -DTrilinos_TEST_CATEGORIES:STRING=BASIC \
  -DTrilinos_DISABLE_ENABLED_FORWARD_DEP_PACKAGES=ON \
  -DTrilinos_ENABLE_ALL_OPTIONAL_PACKAGES:BOOL=ON \
  -DTrilinos_ENABLE_ALL_FORWARD_DEP_PACKAGES:BOOL=ON \
  -DParMETIS_INCLUDE_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/parmetis/4.0.3/gcc/4.9.3/openmpi/1.8.7/64bit_parallel/include \
  -DParMETIS_LIBRARY_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/parmetis/4.0.3/gcc/4.9.3/openmpi/1.8.7/64bit_parallel/lib \
  -DBoostLib_INCLUDE_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/boost/1.63.0/gcc/4.9.3/base/include \
  -DBoostLib_LIBRARY_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/boost/1.63.0/gcc/4.9.3/base/lib \
  -DBoost_INCLUDE_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/boost/1.63.0/gcc/4.9.3/base/include \
  -DHDF5_INCLUDE_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/hdf5/1.8.12/gcc/4.9.3/openmpi/1.8.7/parallel/include \
  -DHDF5_LIBRARY_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/hdf5/1.8.12/gcc/4.9.3/openmpi/1.8.7/parallel/lib \
  -DNetcdf_INCLUDE_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/netcdf/4.3.2/gcc/4.9.3/openmpi/1.8.7/parallel/include \
  -DNetcdf_LIBRARY_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/netcdf/4.3.2/gcc/4.9.3/openmpi/1.8.7/parallel/lib \
  -DTPL_ENABLE_SuperLU:BOOL=ON \
  -DSuperLU_INCLUDE_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/superlu/4.3/gcc/4.9.3/base/include \
  -DSuperLU_LIBRARY_DIRS=/projects/sems/install/rhel6-x86_64/sems/tpl/superlu/4.3/gcc/4.9.3/base/lib \
  -DTPL_ENABLE_Matio:BOOL=OFF \
  /scratch/bmpersc/sierra_integration2/test_runs/Trilinos


Note that this will build much more than you need, but it is a known configuration that shows the issue.

**Comment by bmpersc** [comment]
@wfspotz Here is the module commands I run for my configuration above.

module purge
module load sems-env
module load sems-python/2.7.9
module load sems-cmake/3.5.2
module load sems-gcc/4.9.3
module load sems-openmpi/1.8.7
module load sems-parmetis/4.0.3/64bit_parallel
module load sems-boost/1.63.0/base
module load sems-hdf5/1.8.12/parallel
module load sems-netcdf/4.3.2/parallel
module load sems-superlu/4.3/base
module load sems-yaml_cpp/0.5.3/base

Again, this will have much more than you need for Domi.

**Comment by bartlettroscoe** [comment]
@wfspotz and I pair-programmed the fix for this today and I pushed the fix in the commit 14d9e83.  Therefore, these failures should be gone tomorrows.  I am therefore, putting this in review.

NOTE: These unit tests got bit by undefined behaviour and the lack of debug-mode checking described in #1331.  In a debug-mode build (which the standard CI build uses), it just happened to pass.  But in a non-debug-mode build, the undefined behaviour become apparent.  Also note that valgrind did not report any issue because this was undefined behaviour related to reuse of deleted objects on the stack and not the heap (and valgrind can't see anything on the stack).

**Comment by bartlettroscoe** [comment]
This has been fixed for a long time.  For example, see the all-passing Domi tests on the Clean dashboard at:

* http://testing.sandia.gov/cdash/index.php?project=Trilinos&date=2017-06-01&subproject=Domi

Closing as complete.
